src/tests/test_tpm2.cpp

```diff
@@ -0,0 +1,794 @@
+/*
+* (C) 2024 Jack Lloyd
+* (C) 2024 René Meusel, Amos Treiber Rohde & Schwarz Cybersecurity
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#include "tests.h"
+
+#include <botan/internal/fmt.h>
+#include <botan/internal/loadstor.h>
+#include <botan/internal/stl_util.h>
+
+#if defined(BOTAN_HAS_TPM2)
+   #include <botan/internal/tpm2_hash.h>
+
+   #include <botan/pubkey.h>
+   #include <botan/tpm2_key.h>
+   #include <botan/tpm2_rng.h>
+   #include <botan/tpm2_session.h>
+
+   #if defined(BOTAN_HAS_TPM2_RSA_ADAPTER)
+      #include <botan/tpm2_rsa.h>
+   #endif
+
+   #include <tss2/tss2_esys.h>
+#endif
+
+namespace Botan_Tests {
+
+#if defined(BOTAN_HAS_TPM2)
+namespace {
+
+   #if defined(BOTAN_HAS_TPM2_CRYPTO_BACKEND) && defined(BOTAN_TSS2_SUPPORTS_CRYPTO_CALLBACKS)
+constexpr bool crypto_backend_should_be_available = true;
+   #else
+constexpr bool crypto_backend_should_be_available = false;
+   #endif
+
+std::shared_ptr<Botan::TPM2::Context> get_tpm2_context(std::string_view rng_tag) {
+   const auto tcti_name = Test::options().tpm2_tcti_name();
+   if(tcti_name.value() == "disabled") {
+      // skip the test if the special 'disabled' TCTI is configured
+      return {};
+   }
+
+   auto ctx = Botan::TPM2::Context::create(tcti_name, Test::options().tpm2_tcti_conf());
+   if(ctx->vendor() != "SW   TPM" || ctx->manufacturer() != "IBM") {
+      return {};
+   }
+
+   if(ctx->supports_botan_crypto_backend()) {
+      ctx->use_botan_crypto_backend(Test::new_rng(rng_tag));
+   }
+
+   return ctx;
+}
+
+Test::Result bail_out() {
+   Test::Result result("TPM2 test bail out");
+
+   if(Test::options().tpm2_tcti_name() == "disabled") {
+      result.test_note("TPM2 tests are disabled.");
+      return result;
+   } else {
+      result.test_failure("Not sure we're on a simulated TPM2, cautiously refusing any action.");
+      return result;
+   }
+}
+
+bool not_zero_64(std::span<const uint8_t> in) {
+   Botan::BufferSlicer bs(in);
+
+   while(bs.remaining() > 8) {
+      if(Botan::load_be(bs.take<8>()) == 0) {
+         return false;
+      }
+   }
+   // Ignore remaining bytes
+
+   return true;
+}
```

Comment:
There was [an open discussion](https://github.com/randombit/botan/pull/4117#discussion_r1672028429) in the previous PR where Jack pointed out that this could fail randomly with a 1/2^64 probability.